### ISSUE TITLE
Request for clarification and documentation on applying licenses to packages that dynamically link against downstream GPL licensed dependencies

### ISSUE THREAD

**Issue by moorepants** [conda-forge/conda-forge.github.io#1065]
Some conda packages depend on dynamically linking to other conda packages that happen to be GPL licensed. Conda forge then distributes individual binaries for each package separately and the interrelated packages are then dynamically linked on a end user's computer when all the packages are installed and loaded.

The concern/question I have is whether the license of the conda package that is upstream of a dynamically linked GPL package has to be licensed with the GPL or not. And secondly, if so, what should the license include (e.g. licenses for everything that is dynamically linked) and even whether the source code has to be distributed with the conda package somehow.

This question arose here originally:

https://github.com/conda-forge/pyshtools-feedstock/issues/3

where we decided to apply the GPL to the pyshtools package even those the pyshtools source code is BSD licensed. pyshtools dynamically links against FFTW which is licensed "GPL 2 or later".

I then opened some issues on non-GPL licensed upstream dependencies of FFTW to warn those maintainers that they may need to re-license as GPL.

The maintainer of OpenMM, @peastman, responded with some careful explanations in which he argues that the binaries that Conda Forge distributes do not have to be GPL licensed but only GPL compatible.

See those explanations here:

https://github.com/conda-forge/openmm-feedstock/issues/24

My head has exploded trying to think about this, as it is quite confusing. Thus, I think it would be helpful to get a ruling on this and then add some documentation that clearly states the consequences of upstream packages dynamically linking to downstream GPL packages.

**Comment by jakirkham** [comment]
There seem to be a few interrelated questions. Pulling them apart to hopefully move the conversation forward. Also IANAL.

1. What should the license of a library be if it depends on another library that is GPL?
2. Does this change if the dependency is optional?
3. What should the license of a package be?
4. What are the license implications for end users?

AIUI 1 means the library itself should be GPL, but I could be wrong (related discussion suggests this is unclear). AIUI 2 depends on whether the library is linked or not.

With 3, I would propose we just mirror the packaged library's licensing as-is. It's up to the library to correctly reflect this in the license in terms of handling 1 and 2. Our goal is then to make sure the end user has this information available.

With 4, I think this is out-of-scope for conda-forge. It's up to companies, government agencies, institutions, etc. to determine the license implications for them and handle as needed. Should add when a collection of packages is installed onto a user's system, they have all of the licenses and license information available to them to determine next steps.

**Comment by jakirkham** [comment]
cc @conda-forge/core (for more/other thoughts :)

**Comment by beckermr** [comment]
> With 4, I think this is out-of-scope for conda-forge. It's up to companies, government agencies, institutions, etc. to determine the license implications for them and handle as needed. Should add when a collection of packages is installed onto a user's system, they have all of the licenses and license information available to them to determine next steps.

:100: to this. We are not lawyers and should not ever give any impression that we are supplying advice to others.

**Comment by moorepants** [comment]
> What should the license of a library be if it depends on another library that is GPL?

I think that this should be "What should the license of a library be if it depends on *and dynamically links against* another library that is GPL during its build process?"

> I would propose we just mirror the packaged library's licensing as-is. It's up to the library to correctly reflect this in the license in terms of handling 1 and 2.

> With 4, I think this is out-of-scope for conda-forge. It's up to companies, government agencies, institutions, etc. to determine the license implications for them and handle as needed. Should add when a collection of packages is installed onto a user's system, they have all of the licenses and license information available to them to determine next steps.

I'm not sure these are the safest interpretations, because Conda Forge and Anaconda.org (where the binaries are stored) are distributing the binaries. I've long been under the impression that if a person or company or other entity distributes software they have to abide by the license during that act of distributing. Maybe its worth understanding who is it that distributes a conda built binary?

> We are not lawyers and should not ever give any impression that we are supplying advice to others.

These kinds of statements are always made on seemingly every discussion of licenses, but I'm not sure that statements like this are helpful. As a member of society that has laws, we have to interpret the meaning of the laws and choose actions that abide by them to the best of our ability. If every response to license questions is "no comment, we are not lawyers", we'll never make it anywhere because there aren't enough lawyers to go around interpreting life for us (pro-bono or not).

**Comment by moorepants** [comment]
>  With 4, I think this is out-of-scope for conda-forge. It's up to companies, government agencies, institutions, etc. to determine the license implications for them and handle as needed. Should add when a collection of packages is installed onto a user's system, they have all of the licenses and license information available to them to determine next steps.

Yes, what an end user does with the software is not our responsibility. If they use it or distribute it, they have to abide by the licenses and spend their effort interpreting the licenses.

**Comment by beckermr** [comment]
> These kinds of statements are always made on seemingly every discussion of licenses, but I'm not sure that statements like this are helpful. As a member of society that has laws, we have to interpret the meaning of the laws and choose actions that abide by them to the best of our ability. If every response to license questions is "no comment, we are not lawyers", we'll never make it anywhere because there aren't enough lawyers to go around interpreting life for us (pro-bono or not).

The alternative is for us to be giving legal advice which is a huge problem. Hence no comment on how users should interpret and use the license info.

**Comment by moorepants** [comment]
> The alternative is for us to be giving legal advice which is a huge problem.

I've opened this issue as a request for advice on a matter of legality, not a request for "legal advice", i.e. advice from a lawyer (although if lawyers want to chime in that'd be helpful). So I hope that we can discuss this and decide how we should license binary packages built and distributed using Conda Forge's resources.

For me, if I create the build recipe and initiate the action of building the software and distributing it, I want to have some personal confidence that I'm applying and including the licenses correctly in the binary I helped distribute, because I could be held liable for that action.

**Comment by beckermr** [comment]
That's fine. I don't think you'll find anyone here willing to provide advice of this nature, whatever that nature is.

**Comment by beckermr** [comment]
> For me, if I create the build recipe and initiate the action of building the software and distributing it, I want to have some personal confidence that I'm applying and including the licenses correctly in the binary I helped distribute, because I could be held liable for that action.

Agreed 100%. We worry too!

**Comment by moorepants** [comment]
I found a conversation about this same issue in the pyFFTW library's issue tracker. I've read it, there are various opinions, and I don't yet have a good summary, but I'll post it here for informational purposes:

https://github.com/pyFFTW/pyFFTW/issues/229

**Comment by moorepants** [comment]
Also this thread:

https://github.com/xtensor-stack/xtensor-fftw/issues/36

**Comment by moorepants** [comment]
The author of the FFTW license left this comment on the Julia bindings to FFTW:

https://github.com/JuliaMath/FFTW.jl/pull/41#discussion_r141968395

And the resulting message in the README of the Julia bindings is here:

> The FFTW library will be downloaded on versions of Julia where it is no longer distributed as part of Julia. Note that FFTW is licensed under GPLv2 or higher (see its license file), but the bindings to the library in this package, FFTW.jl, are licensed under MIT. This means that code using the FFTW library via the FFTW.jl bindings is subject to FFTW's licensing terms. Code using alternative implementations of the FFTW API, such as MKL's FFTW3 interface are instead subject to the alternative's license. If you distribute a derived or combined work, i.e. a program that links to and is distributed with the FFTW library, then that distribution falls under the terms of the GPL. If you just distribute source code that links to FFTW.jl, and users have to download FFTW or MKL to provide the backend, then the GPL probably doesn't have much effect on you.

So, this seems to indicate if FFTW is downloaded separately from the software that links to it (as is done with conda) then the GPL does not infect the software.

**Comment by moorepants** [comment]
More related issues:

- https://github.com/conda-forge/conda-forge.github.io/issues/209
- https://github.com/conda-forge/pyfftw-feedstock/issues/24

**Comment by peastman** [comment]
> If the package license is already GPL then do no work otherwise the following:

That should be, "If the package license is GPL compatible then do no work."  As long as your license is GPL compatible (MIT, BSD, LGPL...), there's no problem with linking to GPL libraries.

**Comment by bryan-hunt** [comment]
> > If the package license is already GPL then do no work otherwise the following:
> 
> That should be, "If the package license is GPL compatible then do no work." As long as your license is GPL compatible (MIT, BSD, LGPL...), there's no problem with linking to GPL libraries.

That is only true for LGPL and when there is a GPL exception in place for linking (such as the famous linux userspace linking exception). The GPL is explicitly designed to propagate. You can include lesser license like MIT/BSD/LGPL etc code within a GPL project - you can't relicense GPL code to a less restrictive license. Compatibility is not bi-directional - many of the libraries that are GPL do make this explicit as well. See https://www.gnu.org/licenses/gpl-faq.en.html#WhatDoesCompatMean

Any time a lesser license links to a GPL library you need to confirm that the author of that library does not intend for the license to propagate. If there is no explicit declaration that it does not supersede the license of the resulting linked application then you must consider that it does.

Some additional references: https://wikipedia.org/wiki/GPL_linking_exception

**Comment by bryan-hunt** [comment]
> That's not correct. You are free to distribute your own code under any license you want. It doesn't include any GPL code, so there's no problem.
> 
> When your program gets loaded into memory, and links to a separately installed GPL library, that linked executable in memory becomes a "derived product" under the terms of the GPL and must be treated as GPL licensed. Your own library must therefore have a license that permits doing that. This is what's meant by "GPL compatible".
> 
> https://www.gnu.org/licenses/gpl-faq.html#LinkingWithGPL

Yes - under most circumstances the packages we're talking about are compiled libraries and not severable from their dependency. This is why there is the recipe license and the package license - the package license reflecting the source code license rather than the derived work license is a misrepresentation.

This gets weird when talking about python - but when it comes to compiled libraries it ends up being pretty clear.
